### ISSUE TITLE
`no_effect_underscore_binding` fires when ignoring a generic argument

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust-clippy#8300]
### Summary

When accepting a `T: IntoIterator` argument, `no_effect_underscore_binding` requires `.into_iter()` be called on the argument when assigning to an `_` prefixed variable.

### Lint Name

no_effect_underscore_binding

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
#![warn(clippy::all)]
#![warn(clippy::pedantic)]

/// # Errors
/// not implemented
pub fn iter_good<T>(args: T) -> Result<String, &'static str>
where
    T: IntoIterator<Item = String>,
{
    let _ignored_while_unimplemented = args.into_iter();
    Err("good")
}

/// # Errors
/// not implemented
pub fn iter_bad<T>(args: T) -> Result<String, &'static str>
where
    T: IntoIterator<Item = String>,
{
    let _ignored_while_unimplemented = args;
    Err("bad")
}
```

I expected to see this happen: no warnings

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ffd5b85bf0c41df91d23f7eb9e1eea9

### Version

```text
playground rust 1.58.0

Clippy 0.1.60 (2022-01-15 ec4bcaa)
```

**Comment by oxalica** [comment]
The lint also falsely triggers when the binding **actually has effects** on drop ordering, like to move a generic guard into an `async` block.
```rust
#![warn(clippy::pedantic)]

fn work(_: impl std::future::Future<Output = ()>) { todo!() }

pub fn foo<T>(guard: T) {
    let fut = async move {
        let _guard = guard;
        // work...
    };
    work(fut);
}
```
